test/cpp/api/sequential.cpp

```diff
@@ -32,20 +32,43 @@ TEST_F(SequentialTest, ConstructsFromSharedPointer) {
   Sequential sequential(
       std::make_shared<M>(1), std::make_shared<M>(2), std::make_shared<M>(3));
   ASSERT_EQ(sequential->size(), 3);
+
+  Sequential sequential_named(named_submodules({
+    {"m1", std::make_shared<M>(1)},
+    {std::string("m2"), std::make_shared<M>(2)},
+    {"m3", std::make_shared<M>(3)}
+  }));
+  ASSERT_EQ(sequential->size(), 3);
 }
 
 TEST_F(SequentialTest, ConstructsFromConcreteType) {
   struct M : torch::nn::Module {
     explicit M(int value_) : value(value_) {}
+    M(const M& other) : torch::nn::Module(other) {
+      // NOTE: The current implementation expects the module to be copied once
+      // when it's passed into `std::make_shared<T>()`.
+      // TODO: Find a way to avoid copying, and then delete the copy constructor.
+      AT_ASSERT(copy_count == 0);
```

Comment:
this always has to be true, right?

Reply:
yes this is only true when the copy constructor is called the first time, since we increment `copy_count` after this line. When the problem mentioned at https://github.com/pytorch/pytorch/issues/17879 is fixed, we will be able to remove the copy constructor completely and don't need to do this check anymore.

Reply:
how can the copy constructor of an object be called multiple times?

Reply:
It was an oversight, I fixed it now with a global variable that stores the copy count.